aptos-move/framework/aptos-framework/sources/transaction_fee.move

```diff
@@ -1,15 +1,127 @@
+            amount: coin::initialize_aggregatable_coin(aptos_framework),
+            proposer: option::none(),
+            burn_percentage,
+        };
+        move_to(aptos_framework, collected_fees);
+    }
+
+    fun is_fees_collection_enabled(): bool {
+        exists<CollectedFeesPerBlock>(@aptos_framework)
+    }
+
+    /// Sets the burn percentage for collected fees to a new value. Should be called by on-chain governance.
+    public fun upgrade_burn_percentage(aptos_framework: &signer, new_burn_percentage: u8) acquires CollectedFeesPerBlock {
+        system_addresses::assert_aptos_framework(aptos_framework);
+        assert!(new_burn_percentage <= 100, error::out_of_range(EINVALID_BURN_PERCENTAGE));
+
+        if (is_fees_collection_enabled()) {
+            // Upgrade has no effect unless fees are being collected.
+            let burn_percentage = &mut borrow_global_mut<CollectedFeesPerBlock>(@aptos_framework).burn_percentage;
+            *burn_percentage = new_burn_percentage
+        }
+    }
+
+    /// Registers the proposer of the block for gas fees collection. This function
+    /// can only be called at the beginning of the block.
+    public(friend) fun register_proposer_for_fee_collection(proposer_addr: address) acquires CollectedFeesPerBlock {
+        if (is_fees_collection_enabled()) {
+            let collected_fees = borrow_global_mut<CollectedFeesPerBlock>(@aptos_framework);
+            let _ = option::swap_or_fill(&mut collected_fees.proposer, proposer_addr);
+        }
+    }
+
+    /// Burns a specified fraction of the coin.
+    fun burn_coin_fraction(coin: &mut Coin<AptosCoin>, burn_percentage: u8) acquires AptosCoinCapabilities {
+        assert!(burn_percentage <= 100, error::out_of_range(EINVALID_BURN_PERCENTAGE));
+
+        let collected_amount = coin::value(coin);
+        let amount_to_burn = (burn_percentage as u64) * collected_amount / 100;
+        if (amount_to_burn > 0) {
+            let coin_to_burn = coin::extract(coin, amount_to_burn);
+            coin::burn(
+                coin_to_burn,
+                &borrow_global<AptosCoinCapabilities>(@aptos_framework).burn_cap,
+            );
+        }
+    }
+
+    /// Calculates the fee which should be distributed to the block proposer at the
+    /// end of an epoch, and records it in the system. This function can only be called
+    /// at the beginning of the block or during reconfiguration.
+    public(friend) fun process_collected_fees() acquires AptosCoinCapabilities, CollectedFeesPerBlock {
+        if (!is_fees_collection_enabled()) {
+            return
+        };
+        let collected_fees = borrow_global_mut<CollectedFeesPerBlock>(@aptos_framework);
+
+        // If there are no collected fees, do nothing.
+        if (coin::is_aggregatable_coin_zero(&collected_fees.amount)) {
+            return
+        };
+
+        // Otherwise get the collected fee, and check if it can distributed later.
```

Comment:
I remember there was a comment on unsetting the proposer after draining and distributing the collected fees. What happened to that?

Reply:
I used `option::extract` explicitly, will add the comment there as well.